task/dump-agent/src/main.rs

```diff
@@ -0,0 +1,235 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Dump Agent
+
+#![no_std]
+#![no_main]
+
+use core::mem::size_of;
+use dump_agent_api::*;
+use idol_runtime::RequestError;
+use static_assertions::const_assert;
+use userlib::*;
+
+//
+// Our DUMP_READ_SIZE must be an even power of 2 -- and practically speaking
+// cannot be more than 1K
+//
+const_assert!(DUMP_READ_SIZE & (DUMP_READ_SIZE - 1) == 0);
+const_assert!(DUMP_READ_SIZE <= 1024);
+
+struct ServerImpl {
+    areas: [DumpArea; 3],
+}
+
+task_slot!(SPROT, sprot);
+
+impl ServerImpl {
+    fn area(&self, index: usize, offset: u32) -> Result<&[u8], DumpAgentError> {
+        if index >= self.areas.len() {
+            Err(DumpAgentError::InvalidArea)
+        } else {
+            let area = self.areas[index];
+
+            if offset < area.length {
+                let addr = (area.address + offset) as *const u8;
+                let len = (area.length - offset) as usize;
+
+                Ok(unsafe { core::slice::from_raw_parts(addr, len) })
```

Comment:
Would that only come into play if we were operating on this programmatically at all? (That is, the level of initialization doesn't matter -- the value is being shunted from one spot to another.)  And it feels like `read_volatile` has the same constraints with respect to initialization? (That is, "src must point to a properly initialized value of type T.")  Again, happy to make the change, just not sure that it meaningful changes anything for this code...

Reply:
As far as I can tell, merely constructing such a slice is UB; good point about `read_volatile` having the same constraints.

What about only returning a slice that contains valid data from `area(...)`?

```diff
diff --git a/app/demo-stm32h7-nucleo/app-h753.toml b/app/demo-stm32h7-nucleo/app-h753.toml
index 4441ddf0..de427f3f 100644
--- a/task/dump-agent/src/main.rs
+++ b/task/dump-agent/src/main.rs
@@ -33,9 +34,12 @@ impl ServerImpl {
         } else {
             let area = self.areas[index];
 
-            if offset < area.length {
+            let header = area.address as *mut DumpAreaHeader;
+            let written = unsafe { (*header).written };
+
+            if offset < written {
                 let addr = (area.address + offset) as *const u8;
-                let len = (area.length - offset) as usize;
+                let len = (written - offset) as usize;
 
                 Ok(unsafe { core::slice::from_raw_parts(addr, len) })
             } else {
@@ -176,8 +189,8 @@ impl idl::InOrderDumpAgentImpl for ServerImpl {
 
         let area = self.area(index as usize, offset)?;
 
-        for ndx in 0..rval.len() {
-            rval[ndx] = area[ndx];
+        for (rval, ival) in rval.iter_mut().zip(area.iter()) {
+            *rval = *ival;
         }
 
         Ok(rval)
```

(tested on a Gimletlet with the patch below to enable running without RoT)

Reply:
Do we maybe need `read_volatile()` _in addition to_ limiting the reads to only those bytes that have been initialized? We're reading from an area of memory that is modified outside the knowledge of the running "process" (or really, outside the knowledge of the running processor!). If I take dump 1, call `read_dump()`, take dump 2, and call `read_dump()` again, could the compiler conceivably give me back dump 1's data under the assumption that because I'm reading from a `&[u8]` it couldn't have changed? I'm not sure it's _actually_ possible for the compiler to emit code that would do that (where would it save the original data), but maybe it's theoretically possible in terms of unsafe and the [Rust abstract machine](https://www.ralfj.de/blog/2019/07/14/uninit.html)?

All that said, I think I'm proposing something like this (note: NOT tested, entirely possible there are errors):

```rust
    fn read_dump(
        &mut self,
        _msg: &RecvMessage,
        index: u8,
        offset: u32,
    ) -> Result<[u8; DUMP_READ_SIZE], RequestError<DumpAgentError>> {
        let mut rval = [0u8; DUMP_READ_SIZE];

        if offset & ((rval.len() as u32) - 1) != 0 {
            return Err(DumpAgentError::UnalignedOffset.into());
        }

        let area = self
            .areas
            .get(index as usize)
            .ok_or(DumpAgentError::InvalidArea)?;

        let written = unsafe {
            let header = area.address as *mut DumpAreaHeader;
            core::ptr::read_volatile(header).written
        };

        if written > offset {
            let initialized = written - offset;
            let base = area.address as *const u8;
            let base = unsafe { base.add(offset as usize) };

            for i in 0..usize::min(initialized as usize, DUMP_READ_SIZE) {
                rval[i] = unsafe { core::ptr::read_volatile(base.add(i)) };
            }
        }

        Ok(rval)
    }
```

and removing the slice-returning `area()` function altogether.

Reply:
Implemented this, modified to return an error on an out-of-bounds offset. Thank you!